### DISCUSSION TITLE
Internationalization Support

### DISCUSSION THREAD

**Discussion by Shirley-yhm** [o3de/o3de#14287] [General]
We have added Chinese translations for editor and gems in O3DE using the mechanism of QT's internationalization support. And now, we'd like to contribute it to the community. Looking forward to your feedback, thanks.

### Implementation of Translation
1. Use QObject::tr() to wrap strings, like QObject::tr("Hello").
2. Use lupdate in QT to extract strings that need to be translated into files with .ts suffix.
3. Use Linguist Tool in QT to translate the strings in translation files.
4. Use lrelease in QT to compile the translation files with .qm suffix.

### Implementation in QWidgets
 The implementation is just like what has been done in O3DE.
1. Set translators in /Code/Editor/Core/QtEditorApplication.cpp for Editor, Asset Processor and Gems.
2. Load translation module when initialize UI.
3. Unload translation module when exit UI.

### Others
1. For strings not in subclasses of QObject, we can use QT_TRANSLATE_NOOP("Reflect", x) instead of QObject::tr(). Then use QCoreApplication::translate() at where the strings are actually used to translate them.
2. We also need to add a language setting in Editor Settings.

**Comment by AMZN-daimini** [comment]
Hello @Shirley-yhm,
This is great to see! I would be happy to help getting this contribution submitted into the development branch.
Once your contributions are in, I can also help out with adding the necessary UI in the Editor Settings.

How hard is it to produce the translation files and compile them? It would be great if we could also provide documentation for the community to provide more localization files in other languages over time. I would be happy to work on an Italian translation myself when this comes to fruition :)

**Reply by Shirley-yhm** [reply]
Other languages can be added in this mechanism easily after we have tagged all strings that need to be translated in source code. Language will be distinguished ​​by user's setting and filename suffix of translation files. I can provide related documentation later :)

So, what should I do then? Should I write an RFC document to describe it in more detail first?

**Reply by Shirley-yhm** [reply]
How hard is it to produce the translation files and compile them? —— 
1. Tag all strings to be translated in source code.
2. Translation files will be generated for different languages using cmake.
3. Add translations in different languages.

Step 1 and step 2 requires huge workload. We plan to translate Editor and Asset Processor first. Translation for Gems will be added later. Since we have done this work on an old branch, it will progress smoothlier than expected.

**Reply by AMZN-daimini** [reply]
> So, what should I do then? Should I write an RFC document to describe it in more detail first?

Yes, I think an RFC would be helpful
https://github.com/o3de/sig-content/issues/new?assignees=&labels=rfc-feature&template=rfc-feature.md&title=Proposed+RFC+Feature+%3Ddescription%3D

For everything that is handled by Qt/QLinguist under the hood, I would recommend just linking to its documentation. So you don't have to spend a lot of time detailing the whole architecture, just the parts that are relevant (string tagging, translation file format, loading the file... that's about it?)

The portion that would be most interesting would be detailing the different parts of the work so that it's clear what this entails. So from what I can gather:

1. Incremental PRs to tag all strings to translate (Plan to start from the Editor and Asset Processor and go from there)
2. Providing the back-end code to load the translation files at startup and unload them at teardown
3. Work to hook that up to the UI
4. Follow-up documentation work to allow the community to contribute

If it's possible to split up the work that was already done in your branch to separate PRs, it could make sense to divide them by those points above as well (even just 2 PRs for points 1 and 2 separately would be great). That will help reviewers understand what's going on and get more meaningful feedback.
If you want, you can have both the RFC and PRs up in parallel to streamline the process, but it's up to you :)

**Reply by Shirley-yhm** [reply]
Thanks a lot! I really appreciate your support. I will start this work as soon as possible :)

**Reply by Shirley-yhm** [reply]
There is another question... We also have done some work to support Chinese characters in files' name and path. Unfortunately, two issues remain unresolved. I wonder if you have any suggestion on them.

- issue1. It only supports English and Chinese currently. If we want to support other languages, maybe we should use buffer to store strings instead of char*, which may affect the original design in O3DE. Another way is to add special handling for each language, which may not be a good idea. 
  - Do you have any suggestion for this problem? Or maybe we can defer dealing with other languages until a consensus is reached between us.
- issue2. It only supports Windows platform currently.
  - Can we just support Chinese on Windows platform now? Supports on other platforms may be added later when we have a suitable solution to deal with it.

**Reply by AMZN-daimini** [reply]
> * issue1. It only supports English and Chinese currently. If we want to support other languages, maybe we should use buffer to store strings instead of char*, which may affect the original design in O3DE. Another way is to add special handling for each language, which may not be a good idea.

This could likely be a follow-up task that doesn't hold the progress of the initiative. You could probably write a separate issue to sig-content and then I can send it to the teams handling asset processing to try and investigate what the limitations are and what can be done.

> * issue2. It only supports Windows platform currently.
>   * Can we just support Chinese on Windows platform now? Supports on other platforms may be added later when we have a suitable solution to deal with it.

Similarly, what are the technical issues that limit it to Windows only? It could make sense to focus on Windows only for this RFC if we have a viable path to support other OSs as a follow-up. If instead the limitations are insurmountable, then maybe it could make sense to look for alternative solutions.

**Reply by Shirley-yhm** [reply]
Sorry to reply you so late. I'm afraid I still need some discussions in my team. I will finish the RFC of translation this week first.

**Reply by kberg0** [reply]
> * issue1. It only supports English and Chinese currently. If we want to support other languages, maybe we should use buffer to store strings instead of char*, which may affect the original design in O3DE. Another way is to add special handling for each language, which may not be a good idea.

I would recommend we create an RFC to make/treat all char* instances as UTF-8, and then work through the engine identifying areas that are non-compliant and correct them. AZ::Path and AZ::StringUtils and some other areas will probably require some work, but this would effectively make the engine language agnostic and allow it to support any character set.

**Reply by Shirley-yhm** [reply]
I have proposed the RFC about internationalization support in https://github.com/o3de/sig-content/issues/114. If you have any questions or suggestions, please let me know :) If it goes well, we expect to make the first PR this week. @AMZN-daimini 

As for the question about how to support characters in other languages, I'm afraid my team don't have time to deal with this right now :( And we won't commit the relevant code this time. Should I write an issue to describe this problem first? If so, which repo is more appropriate to put this issue? Thanks. @kberg-amzn

**Reply by kberg0** [reply]
Hi Shirley, the suggestion was more for things like filenames, asset names, logging strings, warnings, console output, that might go through AzCore and get treated like regular ascii text. I'm quite certain not all of AzCore is UTF-8 compliant, which will likely result in weird issues over time. On the display side, most rendering toolkits, including Qt, already support extended character sets, so having the stringId / translated display string you've proposed will work great. I don't think we would need to assign any team to the work immediately; just have the issue in place so the community can see localization is still a work in progress.
